drivers/sensor/hx711/hx711.c

```diff
@@ -0,0 +1,528 @@
+/*
+ * Copyright (c) 2020 George Gkinis
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT avia_hx711
+#define HX711_NODE DT_NODELABEL(hx711)
+
+#include <device.h>
+#include <drivers/gpio.h>
+#include <sys/byteorder.h>
+#include <drivers/sensor.h>
+#include <devicetree.h>
+#include <string.h>
+#include <zephyr.h>
+#include <logging/log.h>
+
+#ifdef CONFIG_HX711_DISABLE_INTERRUPTS_WHILE_POLLING
+#include <irq.h>
+#endif
+
+#include "hx711.h"
+
+LOG_MODULE_REGISTER(HX711, CONFIG_LOG_DEFAULT_LEVEL);
+
+static struct hx711_data hx711_data = {
+		.reading = 0,
+		.offset = CONFIG_HX711_OFFSET,
+		.slope = {
+				.val1 = CONFIG_HX711_SLOPE_INTEGER,
+				.val2 = CONFIG_HX711_SLOPE_DECIMAL,
+		},
+		.gain = CONFIG_HX711_GAIN,
+		.rate = CONFIG_HX711_SAMPLING_RATE
+};
+
+static const struct hx711_config hx711_config = {
+		.dout_pin = DT_INST_GPIO_PIN(0, dout_gpios),
+		.dout_ctrl = DT_INST_GPIO_LABEL(0, dout_gpios),
+		.dout_flags = DT_INST_GPIO_FLAGS(0, dout_gpios),
+
+		.sck_pin = DT_INST_GPIO_PIN(0, sck_gpios),
+		.sck_ctrl = DT_INST_GPIO_LABEL(0, sck_gpios),
+		.sck_flags = DT_INST_GPIO_FLAGS(0, sck_gpios),
+
+#if DT_NODE_HAS_PROP(HX711_NODE, rate_gpios)
+		.rate_pin = DT_INST_GPIO_PIN(0, rate_gpios),
+		.rate_ctrl = DT_INST_GPIO_LABEL(0, rate_gpios),
+		.rate_flags = DT_INST_GPIO_FLAGS(0, rate_gpios),
+#endif
+};
+
+/**
+ * @brief Check if HX711 data is ready
+ *
+ * @param data Pointer to the hx711 driver data structure
+ *
+ * @retval true when data is available,
+ * @retval false if data is unavailable
+ */
+
+static bool hx711_is_ready(struct hx711_data *data)
+{
+	return !gpio_pin_get(data->dout_gpio, hx711_config.dout_pin);
+}
+
+/**
+ * @brief Wait until HX711 data is ready
+ *
+ * @param data Pointer to the hx711 driver data structure
+ * @param usec Usecs to wait before returning.
+ *
+ * @retval 0 When data is available,
+ * @retval -EACCES error if usec's are elapsed with no data available.
+ */
+static int hx711_wait_until_ready(struct hx711_data *data, uint32_t usec)
+{
+	while (!hx711_is_ready(data) && usec >= 0) {
+		k_busy_wait(1);
```

Comment:
You're definitely right.
Done.

Reply:
I tried **k_sleep(K_USEC(1));** but then the thread seems to become unresponsive and slow.
No clue why this is. I suspect I have to manually wake up the thread somehow.

I changed it back to **k_busy_wait(1);** and the code flies.

As mentioned [here ](https://docs.zephyrproject.org/1.9.0/kernel/threads/scheduling.html#busy-waiting) if the wait period is small there is sometimes no justification for context switch to take place.

**k_sleep** could be used in conjunction with the **falling edge** of the **DOUT** pin to wake the thread up through an **GPIO interrupt trigger**.
It would be tidy and snappy, but for now It seems OK.

Unless the **1usec** delay seems too big to busy wait I would rather start using the library for my application.

Reply:
I changed this bit to use k_sleep(K_USEC(10)) as you suggested.

The time needed for HX711 to power up is 566ms. I use usec == 600000 now.
It is also the maximum amount of time observed for the HX711 to be ready to provide  a reading.

If that time is elapsed -EACCES is returned.